component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java

```diff
@@ -72,185 +73,322 @@
+            Date expirationTime = claimsSet.getExpirationTime();
+            String jti = claimsSet.getJWTID();
+            Date nbf = claimsSet.getNotBeforeTime();
+            Date issuedAtTime = claimsSet.getIssueTime();
+            long currentTimeInMillis = System.currentTimeMillis();
+            long timeStampSkewMillis = OAuthServerConfiguration.getInstance().getTimeStampSkewInSeconds() * 1000;
+            OAuthAppDO oAuthAppDO = getOAuthAppDO(jwtSubject);
+            String tenantDomain = oAuthAppDO.getUser().getTenantDomain();
+
+            if (!isValidSignature(signedJWT, tenantDomain, jwtSubject)) {
+                return false;
+            }
+            //Validate issuer and subject.
+            if (!validateIssuer(jwtIssuer, oAuthAppDO) || !validateSubject(jwtSubject)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Invalid jwtIssuer: " + jwtIssuer + " or jwtSubject: " + jwtSubject + " is found in the assertion." +
+                            " Expected value for issuer and subject is: " + oAuthAppDO.getOauthConsumerKey());
+                }
+                return false;
+            }
+
+            //Get audience.
+            String validAud = getValidAudience(tenantDomain);
+            long expTime = 0;
+            long issuedTime = 0;
+            if (expirationTime != null) {
+                expTime = expirationTime.getTime();
+            }
+            if (issuedAtTime != null) {
+                issuedTime = issuedAtTime.getTime();
+            }
+
+            //Validate mandotory feilds, audience, nbf,exp time, jti.
+            if (!validateMandatoryFeilds(mandatoryClaims, claimsSet) || !validateAudience(validAud, audience) ||
+                    !validateJWTWithExpTime(expirationTime, currentTimeInMillis, timeStampSkewMillis) ||
+                    !validateNotBeforeClaim(currentTimeInMillis, timeStampSkewMillis, nbf) ||
+                    !validateJTI(signedJWT, jti, currentTimeInMillis, timeStampSkewMillis, expTime, issuedTime) ||
+                    !validateAgeOfTheToken(issuedAtTime, currentTimeInMillis, timeStampSkewMillis)) {
+                return false;
+            }
+
+            return true;
+
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException(e.getMessage(), OAuth2ErrorCodes.INVALID_REQUEST);
         }
+    }
 
-        if (!validateJTI(signedJWT, jti, currentTimeInMillis, timeStampSkewMillis, expirationTime.getTime(),
-                issuedAtTime.getTime())) {
-            return false;
+    private boolean validateMandatoryFeilds(List<String> mandatoryClaims, ReadOnlyJWTClaimsSet claimsSet) throws OAuthClientAuthnException {
+
+        for (String mandatoryClaim : mandatoryClaims) {
+            if (claimsSet.getAllClaims().get(mandatoryClaim) == null) {
+                String errorMessage = "Mandatory field/feilds (Issuer, Subject, Expiration time , JWT ID or " +
```

Comment:
change this debug log to specify which mandatory claim is missing, or specify all the mandatoryClaims defined, rather hadrcoding here.